### PR TITLE
Creating new branch scratch-1. adding QR library and scratch of main bot script

### LINKED ISSUE #1: Create project file structure
We should come up with basic project file structure. I can think of these directories for now:
1. Source - contains source files
2. Test - tests
3. Configs - directory for configs (e.g. telegram bot token). I suppose we can use json files as a template of our config files.

### PR DESCRIPTION
**Pull request by Galaxy13**
Adding library qr_make.py, which creates a QR-code image from text. A…dding scratch of the main script, that interacts with Telegram environment, sets basic commands and logic of bot, receives and sends back a QR-code.

Adding blank file main.py of the main script.
Updating changes of the description of the project in README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
 Python Project with cloud Docker packaging
 
-This bot receives messages from end-user, remembers them, and send you back in fixed time. 
+This bot receives text message from end-user, remembers them, and send you back a QR-code. 
 
 Idea is *cloud based* application, which has to go through **pytest**, with using **Travis CI** for auto building and pulling app to **Docker** container.
 
-Main logic of this bot is in progress, for now. 
+To create the interaction between Telegram and local methods, was used a **pyTelegramBotAPI** 5.3
+https://github.com/eternnoir/pyTelegramBotAPI
```
Comment (Overflow922) [inline README.md:9]: Please add usage section. We need to document application behaviour.

File: main.py [hunk 2]
```diff
@@ -0,0 +1 @@
+### Awaiting tests to push in main.py ###
```
Comment (Overflow922) [inline main.py]: Please don't commit empty files. I think it is better to add this file when tests will be written

File: libraries/qr_make.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import qrcode
+
+def qr_make(web_link, CHAT_ID):
+    qr = qrcode.QRCode(
+    version=6,
+    error_correction=qrcode.constants.ERROR_CORRECT_H,
+    box_size=10,
+    border=4,
+)
```
Comment (Overflow922) [inline libraries/qr_make.py]: Should we use autoformatter tool?

File: libraries/qr_make.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+    qr.add_data(web_link)
+    qr.make(fit=True)
+    img = qr.make_image(fill_color="black",
+                    back_color='white').convert('RGB')
+    img.save('test_' + CHAT_ID + '.png')
```
Comment (Overflow922) [inline libraries/qr_make.py]: I think it is better to split image creation and image saving into different methods.

File: scratches/test_of_qr.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import telebot  # importing Telegram API classes and functions
```
Comment (Overflow922) [inline scratches/test_of_qr.py]: General comment. I'm a bit confused with this file. The name implies that this is the test. But business logic is here. I don't think we should include scratch into pull requests. If you want to show code, it is better to publish then in a separate branch.

File: scratches/test_of_qr.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import libraries.qr_make as qr  # import qr_make lib with qr functions
+import os
+
+# defining global vars
+TOKEN = '2055598237:AAHNAv9osQkVCNqDzbgwclOHMAHY-F01Cb8'
+# list with ids of users, who pressed start button
```
Comment (Overflow922) [inline scratches/test_of_qr.py]: Let's keep configs in a separate file (we discussed it in this issue https://github.com/Galaxy13/telegram_bot-Docker.test/issues/1). Moreover, it is not a good practice to document your code using comments. The main drawback is we need to keep in sync code and comments, which is an easy thing to forget. For instance, if you change `known_users` to dict  you will have to update comment.

**Review by Overflow922 — COMMENTED** [review]
Do we need `scratches/test.png`?

File: libraries/qr_make.py (hunk 4 above)
Comment (Galaxy13) [reply libraries/qr_make.py]: Updated this, by creating two separate methods in lib

File: scratches/test_of_qr.py (hunk 5 above)
Comment (Galaxy13) [reply scratches/test_of_qr.py]: Got it, pull request is closed

File: scratches/test_of_qr.py (hunk 6 above)
Comment (Galaxy13) [reply scratches/test_of_qr.py]: Updated this with adding token.txt

File: main.py (hunk 2 above)
Comment (Galaxy13) [reply main.py]: Done. File was deleted

**Comment by Overflow922** [comment]
You can continue working on this PR. No need to close it. We can reopen it

File: README.md (hunk 1 above)
Comment (Galaxy13) [reply README.md:9]: updated usage section

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+language: python
+
+python:
+  - 2.7
+  - 3.8
+  - 3.9
```
Comment (Overflow922) [inline .travis.yml]: nice test matrix! Do you need to test across python 2?

File: example_token.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+20345458237:ATYNBNAvCjdljvfdghaaNqDzbtyulOFGJOGAIY-4gfhgb8
```
Comment (Overflow922) [inline example_token.txt]: I don't think it is a good idea to keep such files in root folder. Looks like we are using it only for the sake of example.

File: libraries/qr_make.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import qrcode
+
+
+def qr_make(web_link):
+    qr = qrcode.QRCode(
+        version=6,
+        error_correction=qrcode.constants.ERROR_CORRECT_H,
+        box_size=10,
+        border=4,
+    )
```
Comment (Overflow922) [inline libraries/qr_make.py:10]: Any auto-formatter tool for Python that you are aware of?

File: scratches/botTest.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
+import telebot  # importing Telegram API classes and functions
+import libraries.qr_make as qr  # import qr_make lib with qr functions
+from telebot import types
+import os
+import libraries.url as val
+
+# defining global vars
+TOKEN = None
+
+with open('token.txt') as f:
```
Comment (Overflow922) [inline scratches/botTest.py]: I think it is a good idea to refactor this into class. Global variables have side effects, Works for now, but later can shoot you in the foot.

File: .gitignore [hunk 11]
```diff
@@ -122,4 +122,8 @@ venv.bak/
 # mypy
 .mypy_cache/
 .dmypy.json
-dmypy.json
\ No newline at end of file
+dmypy.json
+
+# token of bot
+scratches/token.txt
```
Comment (Overflow922) [inline .gitignore]: I still see this file in PR

File: tests/test_url_validation.py [hunk 12]
```diff
@@ -0,0 +1,4 @@
+from libraries import url as val
+
+def test_url_validation():
```
Comment (Overflow922) [inline tests/test_url_validation.py:4]: Could you pls add negative tests? E.g.
1. if an empty string should is provided validator should return false
It is better to cover corner cases.

**Review by Overflow922 — COMMENTED** [review]
it is much better now!

File: .travis.yml (hunk 7 above)
Comment (Galaxy13) [reply .travis.yml]: After closer look, I admit, project wouldn't work on legacy in this conditon, so tests in Python 2.7 env, removed

File: example_token.txt (hunk 8 above)
Comment (Galaxy13) [reply example_token.txt]: Maybe create folder 'examples' and put it in there?

File: libraries/qr_make.py (hunk 9 above)
Comment (Galaxy13) [reply libraries/qr_make.py:10]: What do you mean by auto formatter? Module, that checks rules of PEP?

File: .gitignore (hunk 11 above)
Comment (Galaxy13) [reply .gitignore]: Yes, I can see it it too. Necessarily to update token and fix .gitignore file

File: example_token.txt (hunk 8 above)
Comment (Overflow922) [reply example_token.txt]: yes, that will do

File: libraries/qr_make.py (hunk 9 above)
Comment (Overflow922) [reply libraries/qr_make.py:10]: Yes, I raised an issue regarding this. https://github.com/Galaxy13/telegram_bot-Docker.test/issues/5

Comment (Overflow922) [reply libraries/qr_make.py:10]: Please don't address that issue in this PR

Comment (Galaxy13) [reply libraries/qr_make.py:10]: For formatting, I am using a built-it Pycharm PEP checker, but for a better test, trying to build tests on the Black pip package

Comment (Galaxy13) [reply libraries/qr_make.py:10]: Added check on PEP format style with Black package.

File: scratches/botTest.py (hunk 10 above)
Comment (Galaxy13) [reply scratches/botTest.py]: I have changed global variable to class Token (commit cfb6c40)

File: tests/test_url_validation.py (hunk 12 above)
Comment (Galaxy13) [reply tests/test_url_validation.py:4]: Added 2 tests in commit ed5e879

**Review by Overflow922 — APPROVED** [review]
